shell/platform/darwin/macos/framework/Source/FlutterEngine.mm

```diff
@@ -703,6 +729,36 @@ - (void)registerViewController:(FlutterViewController*)controller forId:(Flutter
   [controller setUpWithEngine:self viewId:viewId threadSynchronizer:_threadSynchronizer];
   NSAssert(controller.viewId == viewId, @"Failed to assign view ID.");
   [_viewControllers setObject:controller forKey:@(viewId)];
+
+  if (controller.viewLoaded) {
+    [self viewControllerViewDidLoad:controller];
+  }
+}
+
+- (void)viewControllerViewDidLoad:(FlutterViewController*)viewController {
+  __weak FlutterEngine* weakSelf = self;
+  FlutterVSyncWaiter* waiter = [[FlutterVSyncWaiter alloc]
+      initWithDisplayLink:[FlutterDisplayLink displayLinkWithView:viewController.view]
+                    block:^(CFTimeInterval timestamp, CFTimeInterval targetTimestamp,
+                            uintptr_t baton) {
+                      // CAMediaTime and flutter time are both mach_absolute_time.
+                      uint64_t timeNanos = timestamp * 1000000000;
+                      uint64_t targetTimeNanos = targetTimestamp * 1000000000;
+                      FlutterEngine* engine = weakSelf;
+                      if (engine) {
+                        // It is a bit unfortunate that embedder requires OnVSync call on
+                        // main thread just to immediately redispatch it to UI thread.
```

Comment:
nit: `macOS main thread` for clarity since we've got so many thread thread names, and often refer to that thread as the Flutter platform thread.